metricbeat/module/containerd/cpu/metricset.go

```diff
@@ -0,0 +1,249 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cpu
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/v7/metricbeat/module/kubernetes/util"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/helper/prometheus"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+// Metricset for apiserver is a prometheus based metricset
+type metricset struct {
+	mb.BaseMetricSet
+	prometheusClient           prometheus.Prometheus
+	prometheusMappings         *prometheus.MetricsMapping
+	preSystemCpuUsage          float64
+	preContainerCpuTotalUsage  map[string]float64
+	preContainerCpuKernelUsage map[string]float64
+	preContainerCpuUserUsage   map[string]float64
+}
+
+var _ mb.ReportingMetricSetV2Error = (*metricset)(nil)
+
+// getMetricsetFactory as required by` mb.Registry.MustAddMetricSet`
+func getMetricsetFactory(prometheusMappings *prometheus.MetricsMapping) mb.MetricSetFactory {
+	return func(base mb.BaseMetricSet) (mb.MetricSet, error) {
+		pc, err := prometheus.NewPrometheusClient(base)
+		if err != nil {
+			return nil, err
+		}
+		return &metricset{
+			BaseMetricSet:              base,
+			prometheusClient:           pc,
+			prometheusMappings:         prometheusMappings,
+			preSystemCpuUsage:          0.0,
+			preContainerCpuTotalUsage:  map[string]float64{},
+			preContainerCpuKernelUsage: map[string]float64{},
+			preContainerCpuUserUsage:   map[string]float64{},
+		}, nil
+	}
+}
+
+// Fetch gathers information from the containerd and reports events with this information.
+func (m *metricset) Fetch(reporter mb.ReporterV2) error {
+	events, err := m.prometheusClient.GetProcessedMetrics(m.prometheusMappings)
+	if err != nil {
+		return errors.Wrap(err, "error getting metrics")
+	}
+	var systemTotalNs int64
+	perContainerCpus := make(map[string]int)
+	elToDel := -1
+	for i, event := range events {
```

Comment:
The additional percentances calculations can be configured to be enabled/disabled on demand like what is happening in docker module so as to avoid overloading the in case we have too many containers/metrics and we are interested into getting that much detail.

Reply:
Good idea